src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            addHead(newNode);
+        } else {
+            addTail(newNode);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("An incorrect index was entered - " + index);
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            addHead(newNode);
+        } else if (index == size) {
+            addTail(newNode);
+        } else if (index == 0) {
+            Node<T> newHead = new Node<>(head, value, null);
+            head.previous = newHead;
+            head = newHead;
+        } else {
+            Node<T> nodeByIndex = searchNode(index);
+            Node<T> previousNode = nodeByIndex.previous;
+            Node<T> addedMiddleNode = new Node<>(nodeByIndex, value, previousNode);
+            nodeByIndex.previous = addedMiddleNode;
+            previousNode.next = addedMiddleNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return searchNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = searchNode(index);
+        T oldValue = oldNode.value;
+        oldNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> deletedNode = searchNode(index);
+        unlink(deletedNode);
+        size--;
```

Comment:
move `size--` to `unlink()` method to avoid code duplication, check other places in the code please